Regular-Question-Answers/largest-rectangle-in-histogram.cpp

```diff
@@ -0,0 +1,134 @@
+    h = [1]
+    p = [0] 
+------------------------------------------------------------------------------------------------------------
+    i = 2.
+    
+    h = [1, 5]
+    p = [0, 2]
+------------------------------------------------------------------------------------------------------------
+    i = 3.
+    
+    h = [1, 5, 6]
+    p = [0, 2, 3]
+------------------------------------------------------------------------------------------------------------
+    i = 4.
+    
+    area = 6 * (4-3) = 6
+    h = [1, 5]      
+    p = [0, 2]
+    
+    area = 5 * (4-2) = 10
+    h = [1]
+    p = [0]
+
+    last popped element was greater than the newly pushed element, so use the index of the last popped element.
+
+    h = [1, 2]
+    p = [0, 2]
+------------------------------------------------------------------------------------------------------------
+    i = 5
+
+    h = [1, 2, 3]
+    p = [0, 2, 5]
+------------------------------------------------------------------------------------------------------------
+    we need to calculate the areas of the rectangles that can be created with the elements in the stacks.
+    i = 6
+
+    area = 3 * (6-5) = 3
+    h = [1, 2]
+    p = [0, 2]
+
+    area = 2 * (6-2) = 8
+    h = [1]
+    p = [0]
+
+    area = 1 * (6-0) = 6
+------------------------------------------------------------------------------------------------------------
+    answer = 10.
+*/
+
+class Solution {
+public:
+    int largestRectangleArea(vector<int>& heights) {
+        int size = heights.size();
+        stack<int> h;
+        stack<int> p;
+        
+        int ans = 0;
+        h.push(heights[0]); // initialize the stacks.
+        p.push(0);
+
+        for(int i = 1; i < size; i++){
+            int curr_num = heights[i];
+            int top_value = h.top();
+            int top_index = p.top();
+            
+            if(curr_num > top_value){ // if curr_num > top_value it can be a height to be used with the previous ones.
+                h.push(curr_num);     // and also it is a candidate for the next rectangles.
+                p.push(i);
+            }else{
+                int prev_index = top_index; // to keep track the index of the lastly popped element.
+                while(!h.empty() && curr_num <= top_value){
+                    int area = top_value * (i - top_index);
+                    ans = max(ans, area);
+                    h.pop();
+                    p.pop();
+                    prev_index = top_index;
+                    if(h.empty()) break; // not to get runtime error, if we try to reach top of the stack with size = 0, it will be error.
+                    top_value = h.top();
+                    top_index = p.top();
+                }
+                p.push(prev_index); 
+                h.push(curr_num);
+            }
+        }
+        int i = size;
+        while(!h.empty()){ // when our stacks are not empty, there will be rectangles that are need to be checked.

```

Comment:
Çok elzem değil ama şu minvalde bir açıklama daha iyi olabilir : 
After traversing the array and doing area calculations, if the stacks are still not empty then we need to check for the rectangles among the remaining heights.

Reply:
ekledim.